sfs/mono/drivingfunction.py

```diff
@@ -419,61 +765,236 @@ def nfchoa_25d_plane(omega, x0, r0, n=[0, 1, 0], max_order=None, c=None):
     return 2*1j / r0 * d, selection, secondary_source_point(omega, c)
 
 
+def _max_order_circular_harmonics(N, max_order):
+    r"""Compute maximum order of the circular harmonics expansion for
+    2-dimensional HOA.
+
+    Parameters
+    ----------
+    N : int
+        Total number of secondary sources.
+    max_order : int
+        Maximum order of expansion. If None maximum order is calculated.
+
+    Returns
+    -------
+    int
+        Order.
+
+    """
+    return N // 2 if max_order is None else max_order
+
+
 def sdm_2d_line(omega, x0, n0, xs, c=None):
-    """Line source by two-dimensional SDM.
+    r"""Driving function for 2-dimensional SDM for a virtual line source.
+
+    Parameters
+    ----------
+    omega : float
+        Angular frequency of line source.
+    x0 : (N, 3) array_like
+        Sequence of secondary source positions.
+    n0 : (N, 3) array_like
+        Sequence of normal vectors of secondary sources.
+    xs : (3,) array_like
+        Position of line source.
+    c : float, optional
+        Speed of sound.
 
+    Returns
+    -------
+    d : (N,) numpy.ndarray
+        Complex weights of secondary sources.
+    selection : (N,) numpy.ndarray
+        Boolean array containing ``True`` or ``False`` depending on
+        whether the corresponding secondary source is "active" or not.
+    secondary_source_function : callable
+        A function that can be used to create the sound field of a
+        single secondary source.  See `sfs.mono.synthesize()`.
+
+    Notes
+    -----
     The secondary sources have to be located on the x-axis (y0=0).
     Derived from :cite:`Spors2009`, Eq.(9), Eq.(4).
 
+    Examples
+    --------
+    .. plot::
+        :context: close-figs
+
+        array = sfs.array.linear(32, 0.2, orientation=[0, -1, 0])
+        d, selection, secondary_source = sfs.mono.drivingfunction.sdm_2d_line(
+            omega, array.x, array.n, xs)
+        plot(d, selection, secondary_source)
+
     """
     x0 = util.asarray_of_rows(x0)
     n0 = util.asarray_of_rows(n0)
     xs = util.asarray_1d(xs)
     k = util.wavenumber(omega, c)
     ds = x0 - xs
     r = np.linalg.norm(ds, axis=1)
-    return - 1j/2 * k * xs[1] / r * hankel2(1, k * r)
-
+    d = - 1j/2 * k * xs[1] / r * hankel2(1, k * r)
+    selection = util.source_selection_all(len(x0))
+    return d, selection, secondary_source_line(omega, c)
 
 def sdm_2d_plane(omega, x0, n0, n=[0, 1, 0], c=None):
```

Comment:
sfs/mono/drivingfunction.py:841:1: E302 expected 2 blank lines, found 1

Reply:
fixed, sorry, pep8 check was not done yet.